packages/block-editor/src/components/colors-gradients/control.js

```diff
@@ -0,0 +1,160 @@
+
+// translators: first %s: the gradient name or value (e.g. red to green or linear-gradient(135deg,rgba(6,147,227,1) 0%,rgb(155,81,224) 100%)
+const gradientIndicatorAriaLabel = __( '(Gradient: %s)' );
+
+const colorsAndGradientKeys = [ 'colors', 'disableCustomColors', 'gradients', 'disableCustomGradients' ];
+
+function VisualLabel( { colors, gradients, label, currentTab, colorValue, gradientValue } ) {
+	let value, ariaLabel;
+	if ( currentTab === 'color' ) {
+		if ( colorValue ) {
+			value = colorValue;
+			const colorObject = getColorObjectByColorValue( colors, value );
+			const colorName = colorObject && colorObject.name;
+			ariaLabel = sprintf( colorIndicatorAriaLabel, colorName || value );
+		}
+	} else if ( currentTab === 'gradient' && gradientValue ) {
+		value = gradientValue;
+		const gradientObject = __experimentalGetGradientObjectByGradientValue( gradients, value );
+		const gradientName = gradientObject && gradientObject.name;
+		ariaLabel = sprintf( gradientIndicatorAriaLabel, gradientName || value );
+	}
+
+	return (
+		<>
+			{ label }
+			{ !! value && (
+				<ColorIndicator
+					colorValue={ value }
+					aria-label={ ariaLabel }
+				/>
+			) }
+		</>
+	);
+}
+
+function ColorGradientControlInner( {
+	colors,
+	gradients,
+	disableCustomColors,
+	disableCustomGradients,
+	className,
+	label,
+	onColorChange,
+	onGradientChange,
+	colorValue,
+	gradientValue,
+} ) {
+	const canChooseAColor = onColorChange && ( ! isEmpty( colors ) || ! disableCustomColors );
+	const canChooseAGradient = onGradientChange && ( ! isEmpty( gradients ) || ! disableCustomGradients );
+	const [ currentTab, setCurrentTab ] = useState( gradientValue ? 'gradient' : !! canChooseAColor && 'color' );
+
+	if ( ! canChooseAColor && ! canChooseAGradient ) {
+		return null;
+	}
+	return (
+		<BaseControl
+			className={ classnames( 'block-editor-color-gradient-control', className ) }
+		>
+			<BaseControl.VisualLabel>
+				<VisualLabel
+					currentTab={ currentTab }
+					label={ label }
+					colorValue={ colorValue }
+					gradientValue={ gradientValue }
+				/>
+			</BaseControl.VisualLabel>
+			{ canChooseAColor && canChooseAGradient && (
+				<ButtonGroup className="block-editor-color-gradient-control__button-tabs">
+					<Button
+						isLarge
+						isPrimary={ currentTab === 'color' }
+						isSecondary={ currentTab !== 'color' }
+						onClick={ () => ( setCurrentTab( 'color' ) ) }
+					>
+						{ __( 'Solid Color' ) }
+					</Button>
+					<Button
+						isLarge
+						isPrimary={ currentTab === 'gradient' }
+						isSecondary={ currentTab !== 'gradient' }
+						onClick={ () => ( setCurrentTab( 'gradient' ) ) }
+					>
+						{ __( 'Gradient' ) }
+					</Button>
+				</ButtonGroup>
+			) }
+			{ currentTab === 'color' && (
+				<ColorPalette
+					value={ colorValue }
+					onChange={ canChooseAGradient ?
+						( newColor ) => {
+							onColorChange( newColor );
+							onGradientChange();
+						} :
+						onColorChange
```

Comment:
What is the implication of calling or not calling `onColorChange`? Could we make it so that it doesn't matter? I suppose the call to `onGradientChange` with no arguments is a way to unset values, but why can't we just:

```js
<ColorPalette onChange={ ( newColor ) => onColorChange( newColor ); onGradientChange(); } …
```

Reply:
We don't have a guarantee that onGradientChange exists, this component may be used just for color picking without gradients.
With the logic, I used we also avoid generating a new function on each render for some cases.

Reply:
> We don't have a guarantee that onGradientChange exists, this component may be used just for color picking without gradients.

Soon: `onGradientsChange?()` :) 

> With the logic, I used we also avoid generating a new function on each render for some cases.

For sure, but a nice thing with hooks such as `useCallback` is how they allow you to avoid special logic and rely on built-in memoisation.